### ISSUE TITLE
Support - only sending message to oldest subscriber of Observable

### ISSUE THREAD

**Issue by ORESoftware** [ReactiveX/rxjs#2312]
[edit: Ultimately this thread is about functionality that relates to sending messages to a subset of observers, not all observers; that subset could be just one of the observers or even zero, if the developer so intended. Much of this thread may conflate an observer with a subscription, and a Subject with an Observable, even though technically a Subject is an Observable.]

What I am trying to do is only notify the oldest subscriber to the given Observable. The oldest subscriber to an Observable will dynamically change, of course, as the program lives on, and subscribers unsubscribe. _More generally, I would like to have some control over notifying only a subset of the observers to an Observable._ This seems to work for my current use case:   

```js
        let obsClient = this.obsClient = new Subject<any>();

        this.clientStream = Observable.create(sub => {

            const push = Subscriber.create(v => {
               // I assume that zeroeth element holds oldest sub
                if (push === obsClient.observers[0]) {  
                    sub.next(v);
                }
            });

            return obsClient.subscribe(push);
        });

        function onChange(clientCount) {
            obsClient.next({
                time: Date.now(),
                clientCount
            });
        }
```

as you can see we push values onto this.obsClient using onChange, but our program only subscribes to this.clientStream.

**What I really need to verify is - will the oldest subscriber always to be the zeroeth element in the observers array?**

Doubt this info will be in the docs. But this example of only passing items to one of the subscribers would be a nice one to have in any Rx examples section (I would think).

**Comment by kwonoj** [comment]
>  will the oldest subscriber always to be the zeroeth element in the observers array

: No, you shouldn't assume internals of implementation since it is not guaranteed. Due to various reason, internal behavior can change anytime as long as it does not break public interface. I don't think it need to be documented since this is internals of implementation also trying to patch behavior of Observable. You still can create custom behavior to manage subscription if needed.

**Comment by ORESoftware** [comment]
@kwonoj sure that might be true, but it would be nice if there were a stable public API to access the oldest subscriber and only push messages to that subscriber (or some subset of subscribers)

**Comment by kwonoj** [comment]
> it would be nice if there were a stable public API to access the oldest subscriber and only push messages to that subscriber

I'd actually question that part about specific usecases, since it's generally not expected behavior of Observable by default.

**Comment by ORESoftware** [comment]
@kwonoj I really think users need lower level control. If we have 1000 subscribers, it can be very unperformant for them to all be notified. Sometimes we only want a subset of subscribers to be notified of an event. We need that control. 

I have a use case for this now, not just making this up :)

**Comment by david-driscoll** [comment]
I would argue that if you have to divide subscribers, they should be grouped separately and then merged together to create a new observer.

From what I recall the observer interface has no way to direct events to a specific subscriber, and this is by design.  The underlying implementation may change, so your code may break in unexpected ways.

You could `.observeOn(Scheduler.async)` so that observations are taken to the event queue.  Or if some of the downstream consumers don't need the information all the time something like `debounce(Time)` or `audit(Time)` to control the consumers.

Perhaps a little more detail about the specific use case, what is the overall goal of the code you had?

**Comment by ORESoftware** [comment]
@david-driscoll thanks

your point about internal implementation changing is sort of moot, because I am arguing that this internal implementation should be part of a public stable API, but I agree that there may be a better way to do this

so I have a simple queue, which is enqueue and dequeue methods. There is an option called "controlled" which means we will delay enqueueing an item onto the queue until another event happens:

```js
    _enqControlled(lines: string | string[], opts: IEnqueueOpts) {

        opts = opts || {};

        const priority = opts.priority || 1;
        const isShare = opts.isShare !== false;

        lines = _.flattenDeep([lines]);

        let $add = this.init()
            .flatMap(() => {
                return waitForClientCount(this, {timeout: 3000, count: 25, tries: 25})
            })
            .flatMap(() => {
                return acquireLock(this, '<enqControlled>')
                    .flatMap(obj => {
                        return acquireLockRetry(this, obj)
                    })
                    .map(obj =>
                        ({error: obj.error, id: obj.id, opts: opts}))
                    .flatMap(obj => {
                        return appendFile(this, lines, priority)
                            .map(() => obj)
                            .flatMap(obj => releaseLock(this, obj.id))
                    });
            })
            .catch(err => {
                console.error('\n', ' => add / enqueue error => \n', err.stack || err);
                const force = !String(err.stack || err).match(/acquire lock timed out/);
                return releaseLock(this, force);
            })
            .take(1);

        if (isShare) {
            $add = $add.share();
            $add.subscribe();
        }

        return $add;

    }
```

so as you can see, the operative thing here is the function "waitForClientCount", which will controlling teh delay, which looks like this:

```js

export function waitForClientCount(q: QProto, opts: any): Observable<any> {

    opts = opts || {};
    const count = opts.count || 10;
    const timeout = opts.timeout || 3000;
    const tries = opts.tries || 5;
    const diff = opts.diff || 5;

    let index = 0;

    return q.clientStream.bufferCount(count)
        .filter(value => {

            index++;

            const first = value[0];
            const last = value[value.length - 1];

            console.log(first,last);

            if (last.clientCount < 10) {
                console.log(' client count is less than 10.');
                return true;
            }

            if (index >= tries) {
                // we have tried enough, have to let it through
                console.log('try limit is reached must let through.');
                return true;
            }

            // console.log('client diff:',diff);

            if ((first.clientCount - last.clientCount) > diff) {
                // client buildup is now reduced, so we can make more requests
                console.log('count is less than diff, we let through.');
                return true;
            }

            // console.log('condition not met in wait for client count.');
            return false;

        })
        .take(1);
}

```

**Comment by kwonoj** [comment]
I'm still on same boat with @david-driscoll . It is expected you'd separate subscriber high level, instead of try to low level customization to mutate designed behavior of Observable. I disagree `internal implementation should be part of a public stable API`. 

While I can see some point of your usecases, I still don't see reason why it should be handled in such low level after subscription actually happened. If you'd like try to control subscription, it'd be better code actually managed possible request to subscription in high level and manage subscribe / unsubscribe as needed, or managing subscription by separating via grouping or other means by @david-driscoll suggested.

I'm closing this issue for now by this issue answers initial questions regarding current implementation behavior. If you're feeling above behavior still need to be discussed, I'd like to suggest separate issue to discuss. For now there isn't proper forms to follow, but filling out initial in https://github.com/ReactiveX/rxjs/issues/2299 will help to understand background to discuss.

**Comment by ORESoftware** [comment]
I think everyone is wrong about this. Observables fit the observer pattern. Using the observer pattern, we may have many observers/subscribers. Sometimes we do not want to notify all observers/subscribers. Both for performance and for functionality

This functionality needs to be part of a public API. If not RxJS is failing it's users.

RxJS is supposed to be "lower-level" than Promises, for example.

**Comment by ORESoftware** [comment]
Sorry but I feel the need to escalate this issue 

@blesh @staltz @jayphelps 

It would be an extremely useful feature to allow the user to have control over which subscribers get notified, from the side of the Observable. I have had more than 1 use case for this type of functionality.

If there is an alternative to the implementation I described in this thread to achieve the functionality I want, I would be happy to hear it but frankly have no idea what that would be.

**Comment by trxcllnt** [comment]
@ORESoftware on the surface it seems the problem you have is fundamentally both interactive and reactive (somewhat back-pressure-ish). It's difficult to recommend a specific path forward without knowing more about the problem you're trying to solve, but I've been successful coming up with solutions to push-me-pull-you in Rx in the past, so feel free to hit me up on Slack or wherever if you want to go into more detail.

**Comment by david-driscoll** [comment]
It appears what you're going for is something akin to what pausable did in RxJS4, which _has not yet_ been ported to RxJS5, there were reservations around the back-pressure stuff, so that is definitely a case to take up with the team.  @trxcllnt has already offered some help.

I'm going to play devils advocate, do you have an example of any other observer or publish/subscription systems, that let you pick and choose who you notify when an event is fired?

I don't know of one that exists, examples of ones I know that exist, which follow typical publish / subscription patterns are:
* [NodeJS Event Emitter](https://nodejs.org/api/events.html#events_class_eventemitter)
* [JQuery on/off](http://api.jquery.com/on/)
* [Reactive Extensions.Net subject](https://github.com/Reactive-Extensions/Rx.NET/blob/master/Rx.NET/Source/System.Reactive.Interfaces/Reactive/Subjects/ISubject.Multi.cs#L21) (which RxJS is modeled on)

Reading from the [bible itself (Design Patterns)](https://en.wikipedia.org/wiki/Design_Patterns).

> The Observer pattern describes how to establish [relationships between two or more objects]. The key objects in this pattern are **subject** and **observer**.  A subject may have any number of dependent observers.  All observers are notified whenever the subject undergoes a change in state.  In response, each observer will query the subject to synchronize its state with the subject's state.
>
> ...
>
> ## Applicability
> * When a change to one object requires changing others, and you don't know how many objects need to be changed.
> * When an object should be able to notify other objects without making assumptions about who these objects are.  In other words, you don't want these objects tightly coupled.

**Comment by ORESoftware** [comment]
thanks all, @trxcllnt I will try to describe the functionality here but would happy to go into more detail elsewhere so as not to pollute this thread - is it rxjs.slack.com?

basically I have an enqueue method on a queue. I want to control the rate at which items get enqueue, using some other event stream. That other event stream could be dequeue events or basically how overloaded my locking mechanism is. Every time a lock or unlock call happens, I send out an Observable event.

When a enqueue call gets made, currently it becomes just one more subscriber to an Observable. But I only want to enqueue one item at a time! So I just want to send Observable events to the oldest subscriber. And then once that subscriber unsubscribes, I start sending Observable events to the next subscriber etc.

**Comment by jayphelps** [comment]
We can discuss briefly at the core team on Monday.

I may be misunderstanding but it feels like to me this would not be idiomatic RxJS. As mentioned, right now this sort of thing is an internal implementation detail and I'd resist proposals to make it public. My gut tells me you need to use a higher abstraction to achieve effectively the same thing. Your examples were hard for me to follow but it does indeed look like forms of backpressure handling? If you could distill a simple, generic example we might be able to provide some idiomatic solutions. It doesn't need (and probably shouldn't) be code--describe the _effect_ you're trying to achieve rather than how you want to implement it per say. If you want some sort of "current subscriber" stack, that's different than round robin, etc (one of the reasons I'm not following)

**Comment by jayphelps** [comment]
Reopening as the discussion is still going.

**Comment by jayphelps** [comment]
@ORESoftware sorry, just saw you replied with more explanation of what you want to achieve. I'll try to grok when I get some time.

**Comment by ORESoftware** [comment]
np, I am trying to explain it as best I can, it does have to do with backpressure and rate controlling, if you look directly above where it says "use case", that's the shortest explanation I have.

**Comment by trxcllnt** [comment]
@ORESoftware I'm not sure we have a Gitter yet, but I'm in the Reactive-Extensions one @ https://gitter.im/Reactive-Extensions/RxJS

**Comment by staltz** [comment]
I think the whole problem stems from the fact that you started using an Observable **as if** it had multiple Observers attached to it. This is a fundamental design property in RxJS, explained in the documentation http://reactivex.io/rxjs/manual/overview.html#subscribing-to-observables. Only an RxJS **Subject** has multiple Observers attached to it. Observables do not.

Also, this issue seems like a regular question. I would suggest placing it in a chat or in StackOverflow (where the answers can be rewarded points, unlike here). There may be a way of modelling your problem with RxJS, but if you have to propose fundamental changes to the library in order to do that, that smells something is wrong.

**Comment by ORESoftware** [comment]
@staltz thanks, I posted this same question on SO a couple days ago, and then posted here. I'm very unlikely to get a decent answer on SO about this :)

I am not sure if anyone followed the code, at this point we are just dealing with semantics. At the very beginning I guess I meant to say: 

> 
More generally, I would like to have some control over notifying only a subset of the observers to a (-Observable) Subject.

Basically meant to say Subject instead of Observable. (Although I thought that Subject inherited from Observable so is an Observable...)

**Comment by staltz** [comment]
From the docs:

> Observers are just objects with three callbacks

> A Subscription is an object that represents a disposable resource, usually the execution of an Observable.

A Subscriber implements both Observer interface and Subscription interface, but it's a type used mostly internally and shouldn't be necessary externally. 

> Also, I am a bit confused @staltz , it seems clear that we can have more than 1 subscriptions to an Observable, maybe there is a difference between observers and subscriptions that I don't know about:

There are 3 subscriptions of the same Observable, but the Observable does not have a list containing pointers to those three. All in all I recommend reading the docs like a book (from beginning to end) or watching this: https://www.youtube.com/watch?v=uQ1zhJHclvs

**Comment by ORESoftware** [comment]
ok now that we've done some yak shaving  -  I do believe the original issue represented by this thread still stands. My code actually works TMK in multiple use cases. For the record, here are the two implementations that /at  work:


```js

        let obsEnqueue = this.obsEnqueue = new Subject<any>();

         // this piece of code does a round robin, sending unique items to each subscriber/observer
        this.queueStream = Observable.create(obs => {

            const push = Subscriber.create(v => {
                if ((index % obsEnqueue.observers.length) === obsEnqueue.observers.indexOf(push)) {
                    obs.next(v);
                }
            });

            return obsEnqueue.subscribe(push);
        });

        let obsClient = this.obsClient = new Subject<any>();

       // this piece of code only sends data to the oldest subscriber/observer
        this.clientStream = Observable.create(sub => {

            const push = Subscriber.create(v => {

                if (push === obsClient.observers[0]) {
                    sub.next(v);
                }
            });

            return obsClient.subscribe(push);
        });
```

**Comment by kwonoj** [comment]
Closing as inactive, as well as it is out of scope for original design of Rx itself.
